File: ai_services/domain/cnc/contracts.py
from __future__ import annotations

from enum import Enum
from math import isclose
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


Identifier = Annotated[
    str,
    Field(min_length=1, max_length=128, pattern=r"^[A-Za-z0-9][A-Za-z0-9._:-]*$"),
]
Probability = Annotated[float, Field(ge=0.0, le=1.0)]
NonNegativeFloat = Annotated[float, Field(ge=0.0)]
NonNegativeInt = Annotated[int, Field(ge=0)]


class ContractModel(BaseModel):
    model_config = ConfigDict(
        allow_inf_nan=False,
        extra="forbid",
        frozen=True,
        str_strip_whitespace=True,
    )


class ToolAction(str, Enum):
    CONTINUE = "CONTINUE"
    REPLACE = "REPLACE"


class ExecutionOutcome(str, Enum):
    CONTINUED = "CONTINUED"
    REPLACED = "REPLACED"
    WAITING_FOR_SPARE = "WAITING_FOR_SPARE"
    FAILED = "FAILED"


class RiskObjective(str, Enum):
    EXPECTED_COST = "EXPECTED_COST"
    CVAR = "CVAR"


class CuttingCondition(ContractModel):
    condition_id: Identifier
    load_class: Literal["LIGHT", "NOMINAL", "HEAVY"]


class RULDistribution(ContractModel):
    """Discrete RUL distribution consumed by a policy, measured in simulator steps."""

    support_steps: list[NonNegativeInt] = Field(min_length=2)
    probability_mass: list[Probability] = Field(min_length=2)
    survival_beyond_horizon: Probability
    model_version: Identifier

    @model_validator(mode="after")
    def validate_distribution(self) -> RULDistribution:
        if len(self.support_steps) != len(self.probability_mass):
            raise ValueError("support_steps and probability_mass must have equal length")
        if self.support_steps[0] != 0:
            raise ValueError("support_steps must begin at zero")
        if any(right <= left for left, right in zip(self.support_steps, self.support_steps[1:])):
            raise ValueError("support_steps must be strictly increasing")
        total_probability = sum(self.probability_mass) + self.survival_beyond_horizon
        if not isclose(total_probability, 1.0, rel_tol=0.0, abs_tol=1e-6):
            raise ValueError("probability_mass plus survival_beyond_horizon must sum to one")
        return self


class ToolBeliefState(ContractModel):
    """Observable wear and inferred belief only; simulator latent state is forbidden."""

    tool_age_steps: NonNegativeInt
    observed_wear_um: NonNegativeFloat
    posterior_median_wear_um: NonNegativeFloat
    posterior_std_wear_um: NonNegativeFloat
    rul_distribution: RULDistribution


class MachineObservation(ContractModel):
    machine_id: Identifier
    tool_id: Identifier
    job_id: Identifier | None = None
    cutting_condition: CuttingCondition
    tool_state: ToolBeliefState


class SharedInventoryState(ContractModel):
    spares_available: NonNegativeInt
    capacity: NonNegativeInt

    @model_validator(mode="after")
    def validate_capacity(self) -> SharedInventoryState:
        if self.spares_available > self.capacity:
            raise ValueError("spares_available cannot exceed capacity")
        return self


class FleetObservation(ContractModel):
    schema_version: Literal["2.0"] = "2.0"
    observation_id: Identifier
    episode_id: Identifier
    step: NonNegativeInt
    machines: list[MachineObservation] = Field(min_length=1)
    inventory: SharedInventoryState

    @model_validator(mode="after")
    def validate_machine_identity(self) -> FleetObservation:
        machine_ids = [machine.machine_id for machine in self.machines]
        tool_ids = [machine.tool_id for machine in self.machines]
        if len(machine_ids) != len(set(machine_ids)):
            raise ValueError("machine_id values must be unique within an observation")
        if len(tool_ids) != len(set(tool_ids)):
            raise ValueError("active tool_id values must be unique within an observation")
        return self


class MachineAction(ContractModel):
    machine_id: Identifier
    action: ToolAction


class JointAction(ContractModel):
    schema_version: Literal["2.0"] = "2.0"
    observation_id: Identifier
    actions: list[MachineAction] = Field(min_length=1)

    @model_validator(mode="after")
    def validate_unique_agents(self) -> JointAction:
        machine_ids = [item.machine_id for item in self.actions]
        if len(machine_ids) != len(set(machine_ids)):
            raise ValueError("a joint action may contain only one action per machine")
        return self


class CostConfig(ContractModel):
    replacement_cost: NonNegativeFloat
    failure_cost: NonNegativeFloat
    waiting_cost_per_step: NonNegativeFloat
    unused_life_cost_per_step: NonNegativeFloat


class RiskConfig(ContractModel):
    objective: RiskObjective
    cvar_alpha: Annotated[float, Field(gt=0.0, lt=1.0)] = 0.95


class EnvironmentConfig(ContractModel):
    schema_version: Literal["2.0"] = "2.0"
    environment_id: Identifier
    number_of_machines: Annotated[int, Field(ge=1)]
    spare_capacity: NonNegativeInt
    initial_spares: NonNegativeInt
    horizon_steps: Annotated[int, Field(ge=1)]
    failure_threshold_um: Annotated[float, Field(gt=0.0)]
    seed: NonNegativeInt
    costs: CostConfig
    risk: RiskConfig

    @model_validator(mode="after")
    def validate_initial_inventory(self) -> EnvironmentConfig:
        if self.initial_spares > self.spare_capacity:
            raise ValueError("initial_spares cannot exceed spare_capacity")
        return self


class PolicyRecommendation(ContractModel):
    schema_version: Literal["2.0"] = "2.0"
    observation_id: Identifier
    policy_id: Identifier
    policy_version: Identifier
    actions: JointAction
    estimated_expected_cost: NonNegativeFloat | None = None
    estimated_cvar_cost: NonNegativeFloat | None = None

    @model_validator(mode="after")
    def validate_observation_reference(self) -> PolicyRecommendation:
        if self.actions.observation_id != self.observation_id:
            raise ValueError("recommendation and joint action must reference the same observation")
        return self


class MachineStepOutcome(ContractModel):
    machine_id: Identifier
    requested_action: ToolAction
    outcome: ExecutionOutcome
    tool_id_before: Identifier
    tool_id_after: Identifier | None = None
    incurred_cost: NonNegativeFloat

    @model_validator(mode="after")
    def validate_tool_transition(self) -> MachineStepOutcome:
        valid_outcomes = {
            ToolAction.CONTINUE: {ExecutionOutcome.CONTINUED, ExecutionOutcome.FAILED},
            ToolAction.REPLACE: {
                ExecutionOutcome.REPLACED,
                ExecutionOutcome.WAITING_FOR_SPARE,
            },
        }
        if self.outcome not in valid_outcomes[self.requested_action]:
            raise ValueError("outcome is incompatible with requested_action")
        if self.outcome is ExecutionOutcome.REPLACED and self.tool_id_after is None:
            raise ValueError("a replacement outcome requires tool_id_after")
        if self.outcome is ExecutionOutcome.REPLACED and self.tool_id_after == self.tool_id_before:
            raise ValueError("a replacement outcome requires a new tool_id_after")
        if self.outcome is not ExecutionOutcome.REPLACED and self.tool_id_after is not None:
            raise ValueError("tool_id_after is only valid for a replacement outcome")
        return self


class StepResult(ContractModel):
    schema_version: Literal["2.0"] = "2.0"
    observation_id: Identifier
    episode_id: Identifier
    step: NonNegativeInt
    outcomes: list[MachineStepOutcome] = Field(min_length=1)
    inventory_after: SharedInventoryState
    total_cost: NonNegativeFloat
    episode_terminated: bool

    @model_validator(mode="after")
    def validate_unique_machine_outcomes(self) -> StepResult:
        machine_ids = [outcome.machine_id for outcome in self.outcomes]
        if len(machine_ids) != len(set(machine_ids)):
            raise ValueError("a step result may contain only one outcome per machine")
        return self


class EpisodeSummary(ContractModel):
    schema_version: Literal["2.0"] = "2.0"
    episode_id: Identifier
    policy_id: Identifier
    seed: NonNegativeInt
    steps_completed: NonNegativeInt
    total_cost: NonNegativeFloat
    failure_count: NonNegativeInt
    replacement_count: NonNegativeInt
    waiting_steps: NonNegativeInt


class CNCContractCatalog(ContractModel):
    """Schema-export catalog. This wrapper is not itself a wire payload."""

    environment_config: EnvironmentConfig
    fleet_observation: FleetObservation
    joint_action: JointAction
    policy_recommendation: PolicyRecommendation
    step_result: StepResult
    episode_summary: EpisodeSummary

File: backend/src/engine/contracts.js
const { z } = require('zod');

const identifier = z.string().trim().min(1).max(128)
  .regex(/^[A-Za-z0-9][A-Za-z0-9._:-]*$/);
const finiteNumber = z.number().finite();
const nonNegativeNumber = finiteNumber.nonnegative();
const nonNegativeInteger = z.number().int().nonnegative();
const schemaVersion = z.literal('2.0');

const rulDistributionSchema = z.object({
  support_steps: z.array(nonNegativeInteger).min(2),
  probability_mass: z.array(finiteNumber.min(0).max(1)).min(2),
  survival_beyond_horizon: finiteNumber.min(0).max(1),
  model_version: identifier,
}).strict().superRefine((value, ctx) => {
  if (value.support_steps.length !== value.probability_mass.length) {
    ctx.addIssue({ code: 'custom', message: 'RUL support and probability mass must have equal length' });
  }
  if (value.support_steps[0] !== 0
    || value.support_steps.some((step, index) => index > 0 && step <= value.support_steps[index - 1])) {
    ctx.addIssue({ code: 'custom', message: 'RUL support must start at zero and be strictly increasing' });
  }
  const total = value.probability_mass.reduce((sum, probability) => sum + probability, 0)
    + value.survival_beyond_horizon;
  if (Math.abs(total - 1) > 1e-6) {
    ctx.addIssue({ code: 'custom', message: 'RUL probability mass and survival probability must sum to one' });
  }
});

const inventorySchema = z.object({
  spares_available: nonNegativeInteger,
  capacity: nonNegativeInteger,
}).strict().superRefine((value, ctx) => {
  if (value.spares_available > value.capacity) {
    ctx.addIssue({ code: 'custom', message: 'Available spares cannot exceed capacity' });
  }
});

const machineObservationSchema = z.object({
  machine_id: identifier,
  tool_id: identifier,
  job_id: identifier.nullable().optional().default(null),
  cutting_condition: z.object({
    condition_id: identifier,
    load_class: z.enum(['LIGHT', 'NOMINAL', 'HEAVY']),
  }).strict(),
  tool_state: z.object({
    tool_age_steps: nonNegativeInteger,
    observed_wear_um: nonNegativeNumber,
    posterior_median_wear_um: nonNegativeNumber,
    posterior_std_wear_um: nonNegativeNumber,
    rul_distribution: rulDistributionSchema,
  }).strict(),
}).strict();

const fleetObservationSchema = z.object({
  schema_version: schemaVersion,
  observation_id: identifier,
  episode_id: identifier,
  step: nonNegativeInteger,
  machines: z.array(machineObservationSchema).min(1),
  inventory: inventorySchema,
}).strict().superRefine((value, ctx) => {
  const machineIds = value.machines.map((machine) => machine.machine_id);
  const toolIds = value.machines.map((machine) => machine.tool_id);
  if (new Set(machineIds).size !== machineIds.length) {
    ctx.addIssue({ code: 'custom', message: 'Machine identifiers must be unique' });
  }
  if (new Set(toolIds).size !== toolIds.length) {
    ctx.addIssue({ code: 'custom', message: 'Active tool identifiers must be unique' });
  }
});

const machineActionSchema = z.object({
  machine_id: identifier,
  action: z.enum(['CONTINUE', 'REPLACE']),
}).strict();

const jointActionSchema = z.object({
  schema_version: schemaVersion,
  observation_id: identifier,
  actions: z.array(machineActionSchema).min(1),
}).strict().superRefine((value, ctx) => {
  const machineIds = value.actions.map((action) => action.machine_id);
  if (new Set(machineIds).size !== machineIds.length) {
    ctx.addIssue({ code: 'custom', message: 'Joint actions must contain one action per machine' });
  }
});

const policyRecommendationSchema = z.object({
  schema_version: schemaVersion,
  observation_id: identifier,
  policy_id: identifier,
  policy_version: identifier,
  actions: jointActionSchema,
  estimated_expected_cost: nonNegativeNumber.nullable().optional().default(null),
  estimated_cvar_cost: nonNegativeNumber.nullable().optional().default(null),
}).strict().superRefine((value, ctx) => {
  if (value.actions.observation_id !== value.observation_id) {
    ctx.addIssue({ code: 'custom', message: 'Recommendation and actions must reference the same observation' });
  }
});

const machineStepOutcomeSchema = z.object({
  machine_id: identifier,
  requested_action: z.enum(['CONTINUE', 'REPLACE']),
  outcome: z.enum(['CONTINUED', 'REPLACED', 'WAITING_FOR_SPARE', 'FAILED']),
  tool_id_before: identifier,
  tool_id_after: identifier.nullable().optional().default(null),
  incurred_cost: nonNegativeNumber,
}).strict().superRefine((value, ctx) => {
  const validOutcome = value.requested_action === 'CONTINUE'
    ? ['CONTINUED', 'FAILED'].includes(value.outcome)
    : ['REPLACED', 'WAITING_FOR_SPARE'].includes(value.outcome);
  if (!validOutcome) ctx.addIssue({ code: 'custom', message: 'Outcome is incompatible with requested action' });
  if (value.outcome === 'REPLACED' && (!value.tool_id_after || value.tool_id_after === value.tool_id_before)) {
    ctx.addIssue({ code: 'custom', message: 'Replacement must provide a new tool identifier' });
  }
  if (value.outcome !== 'REPLACED' && value.tool_id_after !== null) {
    ctx.addIssue({ code: 'custom', message: 'Only replacement may provide a new tool identifier' });
  }
});

const stepResultSchema = z.object({
  schema_version: schemaVersion,
  observation_id: identifier,
  episode_id: identifier,
  step: nonNegativeInteger,
  outcomes: z.array(machineStepOutcomeSchema).min(1),
  inventory_after: inventorySchema,
  total_cost: nonNegativeNumber,
  episode_terminated: z.boolean(),
}).strict().superRefine((value, ctx) => {
  const machineIds = value.outcomes.map((outcome) => outcome.machine_id);
  if (new Set(machineIds).size !== machineIds.length) {
    ctx.addIssue({ code: 'custom', message: 'Step results must contain one outcome per machine' });
  }
});

const episodeSummarySchema = z.object({
  schema_version: schemaVersion,
  episode_id: identifier,
  policy_id: identifier,
  seed: nonNegativeInteger,
  steps_completed: nonNegativeInteger,
  total_cost: nonNegativeNumber,
  failure_count: nonNegativeInteger,
  replacement_count: nonNegativeInteger,
  waiting_steps: nonNegativeInteger,
}).strict();

const environmentConfigSchema = z.object({
  schema_version: schemaVersion,
  environment_id: identifier,
  number_of_machines: z.number().int().min(1),
  spare_capacity: nonNegativeInteger,
  initial_spares: nonNegativeInteger,
  horizon_steps: z.number().int().min(1),
  failure_threshold_um: finiteNumber.positive(),
  seed: nonNegativeInteger,
  costs: z.object({
    replacement_cost: nonNegativeNumber,
    failure_cost: nonNegativeNumber,
    waiting_cost_per_step: nonNegativeNumber,
    unused_life_cost_per_step: nonNegativeNumber,
  }).strict(),
  risk: z.object({
    objective: z.enum(['EXPECTED_COST', 'CVAR']),
    cvar_alpha: finiteNumber.gt(0).lt(1).optional().default(0.95),
  }).strict(),
}).strict().superRefine((value, ctx) => {
  if (value.initial_spares > value.spare_capacity) {
    ctx.addIssue({ code: 'custom', message: 'Initial spares cannot exceed capacity' });
  }
});

const engineRequestSchema = z.object({
  episodeId: identifier,
  attempt: z.number().int().positive(),
  environmentConfig: environmentConfigSchema,
  policyId: identifier,
  policyVersion: identifier,
  seed: nonNegativeInteger,
}).strict();

const eventEnvelope = (type, payload) => z.object({
  type: z.literal(type),
  payload,
}).strict();

const engineEventSchema = z.discriminatedUnion('type', [
  eventEnvelope('FleetObservation', fleetObservationSchema),
  eventEnvelope('PolicyRecommendation', policyRecommendationSchema),
  eventEnvelope('StepResult', stepResultSchema),
  eventEnvelope('EpisodeSummary', episodeSummarySchema),
]);

class ContractValidationError extends Error {
  constructor(message, issues = []) {
    super(message);
    this.name = 'ContractValidationError';
    this.code = 'ENGINE_EVENT_INVALID';
    this.issues = issues;
  }
}

function validateEngineRequest(value) {
  const result = engineRequestSchema.safeParse(value);
  if (!result.success) throw new ContractValidationError('Invalid engine request', result.error.issues);
  return result.data;
}

function validateEngineEvent(value) {
  const result = engineEventSchema.safeParse(value);
  if (!result.success) throw new ContractValidationError('Invalid CNC contract v2 event', result.error.issues);
  return result.data;
}

module.exports = {
  ContractValidationError,
  engineRequestSchema,
  engineEventSchema,
  validateEngineRequest,
  validateEngineEvent,
};
